ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,269 @@
+        pref BLOB UNIQUE,
+        nref BLOB UNIQUE
+    )
+"""
+
+TKey = TypeVar("TKey")
+TValue = TypeVar("TValue")
+
+
+class Node(NamedTuple):
+    key: bytes
+    value: bytes
+    pref: bytes
+    nref: bytes
+
+
+class LRUSQLDict(MutableMapping[TKey, TValue]):
+    """
+    SQLite3-backed dictionary that implements an LRU cache.
+    """
+
+    def __init__(
+        self,
+        conn: sqlite3.Connection,
+        key_encoder: Callable[[TKey], bytes],
+        key_decoder: Callable[[bytes], TKey],
+        value_encoder: Callable[[TValue], bytes],
+        value_decoder: Callable[[bytes], TValue],
+        cache_size: int = None,
+    ) -> None:
+        self._conn = conn
+        self.cache_size = cache_size
+        self.key_encoder = key_encoder
+        self.key_decoder = key_decoder
+        self.value_encoder = value_encoder
+        self.value_decoder = value_decoder
+
+        self._execute(CREATE_CACHE_QUERY)
+
+    def __iter__(self) -> Iterator[TKey]:
+        with self._conn:
+            for key in self._conn.execute("SELECT key FROM cache"):
+                yield self.key_decoder(first(key))
+
+    def __len__(self) -> int:
+        result = self._fetch_single_query("SELECT COUNT(*) FROM cache;")
+        # ignore b/c mypy cannot interpret the result as an integer
+        return first(result)  # type: ignore
+
+    def __setitem__(self, key: TKey, value: TValue) -> None:
+        # updates LRU cache
+        try:
+            self.__getitem__(key)
+        except KeyError:
+            self._insert_item(key, value)
+        else:
+            self._update_item(key, value)
+
+    def __getitem__(self, key: TKey) -> TValue:
+        # updates LRU cache
+        serialized_key = self.key_encoder(key)
+        lookup_result = self._fetch_single_query(
+            "SELECT value FROM cache WHERE key=?;", (serialized_key,),
+        )
+
+        if not lookup_result:
+            raise KeyError(str(key))
+
+        (value,) = lookup_result
+
+        deserialized_value = self.value_decoder(value)
+
+        # update cache
+        self.__delitem__(key)
+        self._insert_item(key, deserialized_value)
+
+        return deserialized_value
+
+    def __delitem__(self, key: TKey) -> None:
+        serialized_key = self.key_encoder(key)
+        result = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE key=?;", (serialized_key,),
+        )
+
+        if not result:
+            raise KeyError(str(key))
+
+        node = Node(*result)
+
+        # delete key from cache
+        self._execute(
+            "DELETE FROM cache WHERE key=?;", (node.key,),
+        )
+
+        # update any nrefs/prefs in cache
+        nref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE nref=?;", (node.key,),
+        )
+        pref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE pref=?;", (node.key,),
```

Comment:
Looks like you are selecting more than you need to here since you only access the `key` part of the node.  Can you update these queries to only select `key`?